### PR TITLE
Log outputs from custom comparison scripts

### PR DESCRIPTION
**Pull request by yashpatel6**
# Description

Wrap custom comparison script in logger

## Pipeline Run Results

- Case 1
  - Pipeline-demo-pipeline, log: `/hot/software/package/tool-NF-test/Python/development/unreleased/yashpatel-log-custom-script-outputs/log-nftest-20230925T230759Z.log`

# Checklist

- [X] This PR **does *NOT* contain** Protected Health Information [(PHI)](https://ohrpp.research.ucla.edu/hipaa/). A repo may ***need to be deleted*** if such data is uploaded. <br> Disclosing PHI is a ***major problem***[^1] - Even ***a small leak can be costly***[^2].
  
- [X] This PR **does *NOT* contain** germline genetic data[^3], RNA-Seq, DNA methylation, microbiome or other molecular data[^4].

[^1]: [UCLA Health reaches $7.5m settlement over 2015 breach of 4.5m patient records](https://healthitsecurity.com/news/ucla-health-reaches-7.5m-settlement-over-2015-breach-of-4.5m)
[^2]: [The average healthcare data breach costs $2.2 million, despite the majority of breaches releasing fewer than 500 records.](https://www.ponemon.org/local/upload/file/Sixth%20Annual%20Patient%20Privacy%20%26%20Data%20Security%20Report%20FINAL%206.pdf)
[^3]: [Genetic information is considered PHI.](https://www.genome.gov/about-genomics/policy-issues/Privacy#:~:text=In%202013%2C%20as%20required%20by,genetic%20information%20for%20underwriting%20purposes.)
  [Forensic assays can identify patients with as few as 21 SNPs](https://www.sciencedirect.com/science/article/pii/S1525157817305962)
[^4]: [RNA-Seq](https://www.nature.com/articles/ng.2248), [DNA methylation](https://ieeexplore.ieee.org/document/7958619), [microbiome](https://www.pnas.org/doi/pdf/10.1073/pnas.1423854112), or other molecular data can be used to predict genotypes (PHI) and reveal a patient's identity.

- [X] This PR **does *NOT* contain** other non-plain text files, such as: compressed files, images (*e.g.* `.png`, .`jpeg`), `.pdf`, `.RData`, `.xlsx`, `.doc`, `.ppt`, or other output files.

_&emsp; To automatically exclude such files using a [.gitignore](https://docs.github.com/en/get-started/getting-started-with-git/ignoring-files) file, see [here](https://github.com/uclahs-cds/template-base/blob/main/.gitignore) for example._

- [X] I have read the [code review guidelines](https://uclahs-cds.atlassian.net/wiki/spaces/BOUTROSLAB/pages/3187646/Code+Review+Guidelines) and the [code review best practice on GitHub check-list](https://uclahs-cds.atlassian.net/wiki/spaces/BOUTROSLAB/pages/3189956/Code+Review+Best+Practice+on+GitHub+-+Check+List).

- [X] I have set up or verified the `main` branch protection rule following the [github standards](https://uclahs-cds.atlassian.net/wiki/spaces/BOUTROSLAB/pages/3190380/GitHub+Standards#GitHubStandards-Branchprotectionrule) before opening this pull request.

- [X] The name of the branch is meaningful and well formatted following the [standards](https://uclahs-cds.atlassian.net/wiki/spaces/BOUTROSLAB/pages/3189956/Code+Review+Best+Practice+on+GitHub+-+Check+List), using [AD_username (or 5 letters of AD if AD is too long)]-[brief_description_of_branch].
  
- [X] I have added the major changes included in this pull request to the `CHANGELOG.md` under the next release version or unreleased, and updated the date.

### CODE REVIEW

**Review by zhuchcn — COMMENTED** [review]
So this redirects stdout and stderr from custom script to the main output? Where are stdout and stderr writen previously?

Documentation says `selectors` is for IO multiplexing. Does it handle writing from async tasks like nextflow stdout?

**Comment by yashpatel6** [comment]
> So this redirects stdout and stderr from custom script to the main output? Where are stdout and stderr writen previously?

It redirects the custom script outputs through the logging module set up in NFTest; before, the streams were just going to default stdout and stderr and so the messages from the custom scripts didn't end up in the log files generated by NFTest

> Documentation says `selectors` is for IO multiplexing. Does it handle writing from async tasks like nextflow stdout?

It does, @nwiltsie added the real-time logging recently for the Nextflow command as well: https://github.com/uclahs-cds/tool-NFTest/blob/beb8dfa420c284b7f32c009e5321c9651d1bc557/nftest/NFTestCase.py#L115-L145

File: nftest/NFTestAssert.py [hunk 1]
```diff
@@ -49,8 +51,36 @@ def get_assert_method(self) -> Callable:
...
+                        )
+
+                        while process.poll() is None:
+                            events = selector.select()
+                            for key, _ in events:
+                                line = key.fileobj.readline()
+                                if line:
+                                    # The only case in which this won't be true is when
+                                    # the pipe is closed
+                                    self._logger.log(
```
Comment (nwiltsie) [inline nftest/NFTestAssert.py:79]: It might be worthwhile to use a different logger so that these lines are easier to pick out, but that'd be a nicety.

**Comment by zhuchcn** [comment]
Got it that makes a lot of sense! NFtest is currently single-threaded. But I guess we can add multi-threading or multi-processing to it in the future, might be one of our to-dos already.

**Review by nwiltsie — APPROVED** [review]
The key lines from the log look good!

```
2023-09-25 23:08:29,462 - NFTest - DEBUG - Assertion passed
2023-09-25 23:08:29,464 - NFTest - DEBUG - test/custom_comparison.sh /hot/software/package/tool-NF-test/Python/development/unreleased/yashpatel-log-custom-script-outputs/Even-and-odd-output/pipeline-demo-pipeline-1.0.0-rc.1/1000-divisor13/main_workflow/output/filtered.txt /hot/software/pipeline/pipeline-demo-pipeline/Nextflow/development/expected_test_outputs/1000-divisor13/filtered.txt
2023-09-25 23:08:29,499 - NFTest - DEBUG - Last 10 values match
2023-09-25 23:08:29,499 - NFTest - DEBUG - Assertion passed
```
It might be worthwhile to use a different logger so that the lines pop out a little more...
```
2023-09-25 23:08:29,499 - NFTest.comparator - DEBUG - Last 10 values match
2023-09-25 23:08:29,499 - NFTest - DEBUG - Assertion passed
```
... but that's absolutely not a blocker. Nicely done!

File: nftest/NFTestAssert.py (hunk 1 above)
Comment (yashpatel6) [reply nftest/NFTestAssert.py:79]: Makes sense, I've created an issue to track this for now

**Pull request merged**